sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -394,7 +405,7 @@ class Dataset[T] private[sql](
    * @group basic
    * @since 1.6.0
    */
-  def schema: StructType = queryExecution.analyzed.schema
+  def schema: StructType = resolvedTEncoder.schema
```

Comment:
I'm kind of worried about it. We can't guarantee encoder's schema is always same with plan's schema(in this PR we add a project to try to make them consistent, but it can't handle inner field). If they are different, users may select a un-exist column which exists in schema.

cc @marmbrus